### DISCUSSION TITLE
Calculator Assignment

### DISCUSSION THREAD

**Discussion by RivkaBerlinski** [skyvar/.github#4] [General]
Is there an option for us to work in pairs?
We felt that the work was not efficient enough when we worked in threes.

**Comment by bdsoha** [comment]
Yes

**Comment by bdsoha** [comment]
> Is there an option for us to work in pairs? We felt that the work was not efficient enough when we worked in threes.

Usually a discussion starts off with a *description*.
You immediately asked a question...

**Comment by RivkaBerlinski** [comment]
We are debating between C# and JAVA,
Which language is more relevant to our work and in which one will it be easier for us to carry out the task?

@FridaPshitik

**Reply by bdsoha** [reply]
I personally hate `C#` 👿.

But, there is no difference for this task.
One is not more relevant than the other because there is not a lot of *business*.

**Reply by bdsoha** [reply]
If you do use `C#`, there are 3 popular testing frameworks:

- MSTest
- NUnit
- xUnit

For Java, I would use JUnit with Hamcrest

**Comment by esterlev23** [comment]
@esterlev23 @RivkaZishinski 
We are researching DTT and interested in writing in JS with the JEST library.

**Reply by bdsoha** [reply]
> @esterlev23 @RivkaZishinski We are researching DTT and interested in writing in JS with the JEST library.

**Correction:** *TDD*.

Have you written in `js` in previous assignments?
If so, choose a different language:

- Go
- Rust
- Ruby
- Groovy
- Perl
- Java
- C
- C++
- C#
- PHP
- Python
- Haskell
- Scala
- Any flavor of Shell

**Reply by esterlev23** [reply]
The previous assignments were written in Python.
Is there a preference for a language that is less familiar to us?

**Reply by bdsoha** [reply]
I am trying to promote as much **self-learning** as possible....

**Reply by esterlev23** [reply]
We are now finishing a short work day.
We chose to work in JAVA with Junit - new for us!

**Reply by esterlev23** [reply]
We see that we do not have a comfortable working environment with Java and yes there is already a group that started through Docker and it also gives errors.
That's why we chose to work with JS. (In the previous tasks, we did not work with js, and there is also the test add-on here).

**Reply by esterlev23** [reply]
We started working according to TDD rules.
We conducted tests on getting a correct value,
Any value that is not a string returns an error.
We continued with simple tests that received a short expression and returned a correct value (without precedence)
So far all the tests passed.
Now while writing more complex code we discovered the `eval()` function,
Is it ok to use it?

After all, we understood the idea of TDD and working with tests!

**Reply by bdsoha** [reply]
**NEVER USE `eval`**.
It is a considered a last resort and extremity dangerous function.

**Reply by esterlev23** [reply]
We finished writing the code according to the ADD principles.
Just before the PR, we wanted to ask if it is requested to add `DockerFile `and `README.md`

**Reply by bdsoha** [reply]
- Including a `README.md` is considered best-practice.
- A `Dockerfile` is only required when you have a custom environment or setup process. If not, you can include a **`### Run`** section in the readme with a `docker run` command.

**Comment by MiriamRotenshtein** [comment]
@EsterWeiss @MiriamRotenshtein 
We are starting the task and want to write the code in javascript.

**Reply by bdsoha** [reply]
Have you written in `js` in previous assignments?
If so, choose a different language:

**Reply by MiriamRotenshtein** [reply]
We did'nt use js yet.

**Comment by SaraLitman** [comment]
We want to make sure- Is 'Handle computation precedence' include '()'?

**Reply by bdsoha** [reply]
No. But I am not against the implementation as a *bonus* if you have time.

**Comment by chanaSlumianski** [comment]
@ShaindyBorzikowsky 
We chose to perform the task in Java.
Do we need to install the JDK in Docker?

**Reply by bdsoha** [reply]
No. You should be using a Java-based image.

**Reply by chanaSlumianski** [reply]
Of course, we use the image of JAVA:7,
If so, no need to install JDK? Or is it another image?

**Reply by bdsoha** [reply]
`java:7` is over 6 years old.
Try an [`openjdk`](https://hub.docker.com/_/openjdk) variant.

**Comment by chanaSlumianski** [comment]
But according to the link you gave,
this image is officially deprecated

**Reply by bdsoha** [reply]
Good catch, use `adoptopenjdk/openjdk8:alpine-slim` instead.

**Reply by SaraLitman** [reply]
@chananaimen 

We also work on JAVA and changed to the image you recommended above,
We saw that the JUNIT we need to run the test scenarios is in a package called org.junit, we are unable to install it, how can we do it or is there another way to use this package from within the container without installing?

**Reply by bdsoha** [reply]
To install java dependencies, you need to use a `pom.xml` file, along with package manager:
- Ant
- Maven
- Ivy

**I suggest `maven`.**

**Comment by EsterWeiss** [comment]
@MiriamRotenshtein 
Good Morning.
On Thursday we learned about the jest library, 
we started to implement the writing of tests and writing code according to the principles of tdd.

**Comment by chanaSlumianski** [comment]
we tried install maven but mvn command not run.
![image](https://user-images.githubusercontent.com/127835603/232275087-338e186a-254e-4c26-bb0c-2c18b12813d4.png)

**Reply by bdsoha** [reply]
Like all other tasks, **use docker**.

**Reply by chanaSlumianski** [reply]
> Like all other tasks, **use docker**.

Should I run maven in a separate container?
or put it in the project's existing container
@ShaindyBorzikowsky @SaraLitman @chananaimen

**Reply by bdsoha** [reply]
Doesn't matter.

**Comment by ShaindyBorzikowsky** [comment]
@chanaSlumianski @Ester hershberg
our main can't load and run.
we tried all the suggestions of gogle and also open new maven container and creat new java project without anything new .
but nothing help!!!
what can we do ??

**Comment by chanaSlumianski** [comment]
the problem was because auto package from maven.

**Comment by RivkaBerlinski** [comment]
We are stuck with the exception test according to Hamcrest.
We did not find a site that gives clear examples.
Can you help us?

**Reply by bdsoha** [reply]
`assertThrows`

**Reply by RivkaBerlinski** [reply]
We tried to use it but saw that it was not according to Hamcrest.
And this solution also requires the use of `lambda` which is not supported in the `maven` version

**Reply by bdsoha** [reply]
This is a JUnit function, which you are using to run the test suite.

**Comment by RivkaBerlinski** [comment]
We use `maven` image: `adoptopenjdk/maven-openjdk8`
When we run tests that use `lambda` functions
It shows the following error:
![image](https://user-images.githubusercontent.com/127834984/232981769-eaf63069-ba3d-4792-8a4c-260eef62d0f9.png)

**Reply by bdsoha** [reply]
In the future do not post errors as images unless the error is UI related.

**Reply by bdsoha** [reply]
What does your `pom.xml` include?

Try **merging** the following into the file:

```xml
<build>
    <plugins>
        <plugin>
            <groupId>org.apache.maven.plugins</groupId>
            <artifactId>maven-compiler-plugin</artifactId>
            <version>3.3</version>
            <configuration>
                <source>1.8</source>
                <target>1.8</target>
            </configuration>
        </plugin>
    </plugins>
</build>
```

Alternately:

```xml
<properties>
    <maven.compiler.source>1.8</maven.compiler.source>
    <maven.compiler.target>1.8</maven.compiler.target>
</properties>
```

**Comment by RivkaBerlinski** [comment]
We want to clone `branch` from the main and we get this error:
 `unable to access 'https://github.com/skyvar/calculator-assignment.git/': SSL certificate problem: unable to get local issuer certificate`

**Reply by bdsoha** [reply]
Either use `ssh` or ignore ssl verification.

**Reply by RivkaBerlinski** [reply]
Where we use `ssh`?
We do simple `clone` comand in terminal.

**Reply by bdsoha** [reply]
You can clone using `ssh` or `https`.
Such issues should have been encountered on previous assignments, this is very elementary.

**Comment by bdsoha** [comment]
@skyvar/everyone 

This might be a little bit early, but I want to post an intermediary summary.
When approaching this task using *TDD*, you can start *outside-in* or *inside-out* _(both approaches are correct)_.

- **outside-in:** start with the `Calculator` class, and when needed split into smaller helper classes.
- **inside-out:** start with smaller helper classes and move out to the `Calculator` class.

It is also important to always make sure that a single class does 1-thing and **1-thing only**.
As such, the code below follows the *inside-out* approach. Its purpose is to create *tokens* for a provides expression.

<details>
<summary><code>Tokenizer.spec.ts</code></summary>

```typescript
import { Tokenizer } from './Tokenizer'


describe('Tokenizer', () => {
    it('[generate] generates for an empty expression', () => {
        expect(Tokenizer.generate('')).toStrictEqual([])
    })

    it('[generate] generates a token for operands', () => {
        expect(Tokenizer.generate('1')).toStrictEqual(['1'])
        expect(Tokenizer.generate('1.5')).toStrictEqual(['1.5'])
    })

    it('[generate] generates a token for negative operands', () => {
        expect(Tokenizer.generate('-1')).toStrictEqual(['-1'])
        expect(Tokenizer.generate('-1.5')).toStrictEqual(['-1.5'])
    })

    it('[generate] generates a token for operators', () => {
        expect(Tokenizer.generate('+')).toStrictEqual(['+'])
        expect(Tokenizer.generate('-')).toStrictEqual(['-'])
        expect(Tokenizer.generate('*')).toStrictEqual(['*'])
        expect(Tokenizer.generate('/')).toStrictEqual(['/'])
    })

    it('[generate] generates a token for `-` following a operator', () => {
        expect(Tokenizer.generate('1/-1')).toStrictEqual(['1', '/', '-1'])
        expect(Tokenizer.generate('1*-1')).toStrictEqual(['1', '*', '-1'])
    })

    it('[generate] ignores whitespace', () => {
        expect(Tokenizer.generate(' 1 /   -1 ')).toStrictEqual(['1', '/', '-1'])
    })

    it('[generate] ignores spaces found between operands', () => {
        expect(Tokenizer.generate('2 2')).toStrictEqual(['2', '2'])
        expect(Tokenizer.generate('-2 2')).toStrictEqual(['-2', '2'])
    })

    it('[generate] ignores spaces found between operators', () => {
        expect(Tokenizer.generate('2 * *2')).toStrictEqual(['2', '*', '*', '2'])
        expect(Tokenizer.generate('2 -+ 2')).toStrictEqual(['2', '-', '+', '2'])
        expect(Tokenizer.generate('2 +- 2')).toStrictEqual(['2', '+', '-', '2'])
    })
})
```
</details>

<details>
<summary><code>Tokenizer.ts</code></summary>

```typescript
export class Tokenizer {
    protected static PATTERN = /(?:(?<!\d)-)?\d+(?:\.\d+)?|./g

    public static generate(raw: string): string[] {
        // Can be replaces with:
        // `return raw.match(this.PATTERN)?.filter(token => token !== ' ') ?? []`
        
        const matches = raw.match(this.PATTERN)

        if (!matches) {
            return []
        }

        return matches.filter(token => token !== ' ')
    }
}
```
</details>

**Reply by bdsoha** [reply]
@skyvar/everyone I just uploaded a solution to a `solution` branch.
It is important to look at it for learning purposes *(there is no such thing as 1 solution, someone else may provide other solutions)*.

**Comment by esterlev23** [comment]
Our project architecture seems like this:

```text

├── src
│   └── *
└── test
    └── *
```
Where is it correct to insert the `main.js` file and the `router.js` file?

**Reply by bdsoha** [reply]
In `src`. Be sure to separate the *server* code and the *logic* code.

**Comment by chanaSlumianski** [comment]
We added a function to our code and tried to run tests on it,
But suddenly no test works,
There is no error but in practice no test works.
The information I receive:

``Run tests: 0, failures: 0, errors: 0, skipping: 0``

**Reply by bdsoha** [reply]
Did you change the project's folder structure?

**Comment by chanaSlumianski** [comment]
No,
We only added dependencies to pom file.

**Reply by bdsoha** [reply]
Did you rerun any `mvn` commands to download dependencies?

**Comment by bdsoha** [comment]
@skyvar/everyone All teams that have completed the task should move on to working on a UI *(preferably using `React`)* that will interact with their backend.
